Translate: Adding Interactivity (#434)
reactjs/it.react.dev: src/content/learn/adding-interactivity.md

---
title: Aggiungere le Interazioni
---

<Intro>

Alcune cose sullo schermo si aggiornano in risposta all'input dell'utente. Ad esempio, cliccare su una galleria d'immagini modifica l'immagine corrente. In React, i dati che cambiano nel tempo vengono chiamati *state.* Puoi aggiungere lo state a qualsiasi componente e aggiornarlo secondo necessità. In questo capitolo, imparerai a scrivere componenti che gestiscono interazioni, aggiornano il loro state, e visualizzano diversi output nel tempo.

</Intro>

<YouWillLearn isChapter={true}>

* [Come gestire gli eventi originati dall'utente](/learn/responding-to-events)
* [Come far "ricordare" le informazioni ai componenti tramite lo state](/learn/state-a-components-memory)
* [Come React aggiorna la UI in due fasi](/learn/render-and-commit)
* [Perché lo state non si aggiorna subito dopo averlo modificato](/learn/state-as-a-snapshot)
* [Come accodare più aggiornamenti dello state](/learn/queueing-a-series-of-state-updates)
* [Come aggiornare un oggetto nello state](/learn/updating-objects-in-state)
* [Come aggiornare un array nello state](/learn/updating-arrays-in-state)

</YouWillLearn>

> ## Rispondere agli Eventi {/*responding-to-events*/}

Reviewer: ```suggestion
## Rispondere agli eventi {/*responding-to-events*/}
```

> React ti consente di aggiungere degli *event handler* al tuo JSX. Gli event handler sono le tue funzioni personalizzate che vengono chiamate in risposta alle interazioni dell'utente, come il click, il passaggio del mouse, la messa a fuoco delle input e così via.

Reviewer: ```suggestion
React ti consente di aggiungere degli *event handler* al tuo JSX. Gli event handler sono le tue funzioni personalizzate che vengono chiamate in risposta alle interazioni dell'utente, come il click, il passaggio del mouse, la messa a fuoco degli input del form e così via.
```

<LearnMore path="/learn/responding-to-events">

Leggi **[Rispondere agli Eventi](/learn/responding-to-events)** per imparare ad aggiungere gli event handler.

</LearnMore>

> ## State: La Memoria di un Componente {/*state-a-components-memory*/}

Reviewer: ```suggestion
## State: La memoria di un componente {/*state-a-components-memory*/}
```

> Spesso i componenti devono cambiare ciò che c'è sullo schermo in seguito a un'interazione. Scrivere nel form deve aggiornare l'input, cliccare "avanti" su un carosello deve cambiare l'immagine mostrata, cliccare "acquista" inserisce un prodotto nel carrello. I componenti devono "ricordare" le cose: l'attuale valore della input, l'attuale immagine, il carrello. In React, questo specifico tipo di memoria è detto *state.*

Reviewer: ```suggestion
Spesso i componenti devono cambiare ciò che c'è sullo schermo in seguito a un'interazione. Scrivere nel form deve aggiornare l'input, cliccare "avanti" su un carosello deve cambiare l'immagine mostrata, cliccare "acquista" inserisce un prodotto nel carrello. I componenti devono "ricordare" le cose: l'attuale valore dell'input, l'attuale immagine, il carrello. In React, questo specifico tipo di memoria è detto *state.*
```

> Puoi aggiungere lo state a un componente con un hook [`useState`](/reference/react/useState). Gli *Hook* sono speciali funzioni che consentono ai componenti di usare le funzionalità di React (lo state è una di queste). L'hook `useState` consente di dichiarare una variabile state. Prende lo state iniziale e restituisce una coppia di valori: lo state attuale e una funzione *state setter* che consente di aggiornarlo.

Reviewer: ```suggestion
Puoi aggiungere lo state a un componente con un hook [`useState`](/reference/react/useState). Gli *Hook* sono speciali funzioni che consentono ai componenti di usare le funzionalità di React (lo state è una di queste). L'hook `useState` consente di dichiarare una variabile state. Prende lo state iniziale e restituisce una coppia di valori: lo state attuale e una funzione *state setter* che ti consente di aggiornarlo.
```

<LearnMore path="/learn/state-a-components-memory">

Leggi **[State: La Memoria di un Componente](/learn/state-a-components-memory)** per imparare a memorizzare un valore e aggiornarlo in base alle interazioni.

</LearnMore>

> ## Renderizzazione e Commit {/*render-and-commit*/}

Reviewer: ```suggestion
## Renderizzazione e commit {/*render-and-commit*/}
```

Immagina che i tuoi componenti siano cuochi in cucina, che assemblano deliziosi piatti. In questo scenario, React è il cameriere che prende le comande dai clienti e porta loro gli ordini. Questo processo di richiesta e servizio della UI ha tre fasi:

1. **Triggerare** un render (portare la comanda in cucina)
2. **Renderizzare** il componente (preparare l'ordine in cucina)
3. **Committare i cambiamenti** al DOM (portare l'ordine al tavolo)

<IllustrationBlock sequential>
  <Illustration caption="Trigger" alt="React come cameriere in un ristorante, che prende le comande dagli utenti e li porta alla cucina dei componenti." src="/images/docs/illustrations/i_render-and-commit1.png" />
  <Illustration caption="Renderizzazione" alt="Il cuoco Card consegna a React un nuovo componente Card." src="/images/docs/illustrations/i_render-and-commit2.png" />
  <Illustration caption="Commit" alt="React porta la Card al tavolo dell'utente." src="/images/docs/illustrations/i_render-and-commit3.png" />
</IllustrationBlock>

<LearnMore path="/learn/render-and-commit">

Leggi **[Renderizzazione e Commit](/learn/render-and-commit)** per imparare il ciclo di vita di un aggiornamento UI.

</LearnMore>

> ## Lo State come Istantanea {/*state-as-a-snapshot*/}

Reviewer: ```suggestion
## Lo State come istantanea {/*state-as-a-snapshot*/}
```

<LearnMore path="/learn/state-as-a-snapshot">

Leggi **[Lo State come Istantanea](/learn/state-as-a-snapshot)** per imparare perché lo state appare "fisso" e inalterabile dentro gli event handler.

</LearnMore>

> ## Accodare più Aggiornamenti dello State {/*queueing-a-series-of-state-updates*/}

Reviewer: ```suggestion
## Accodare più aggiornamenti dello state {/*queueing-a-series-of-state-updates*/}
```

console.log(score);  // 0
```

> Puoi risolvere questo problema passando una funzione *updater* quando imposti lo state. Nota come sostituire `setScore(score + 1)` con `setScore(s => s + 1)` aggiusta il bottone "+3". Questo consente di accodare più aggiornamenti dello state.

Reviewer: ```suggestion
Puoi risolvere questo problema passando una *funzione updater* quando imposti lo state. Nota come sostituire `setScore(score + 1)` con `setScore(s => s + 1)` aggiusta il pulsante "+3". Questo consente di accodare più aggiornamenti dello state.
```

<LearnMore path="/learn/queueing-a-series-of-state-updates">

Leggi **[Accodare più Aggiornamenti dello State](/learn/queueing-a-series-of-state-updates)** per imparare ad accodare una sequenza di aggiornamenti dello state.

</LearnMore>

> ## Aggiornare gli Oggetti nello State {/*updating-objects-in-state*/}

Reviewer: ```suggestion
## Aggiornare gli oggetti nello state {/*updating-objects-in-state*/}
```

<LearnMore path="/learn/updating-objects-in-state">

Leggi **[Aggiornare gli Oggetti nello State](/learn/updating-objects-in-state)** per imparare ad aggiornare gli oggetti correttamente.

</LearnMore>

> ## Aggiornare gli Array nello State {/*updating-arrays-in-state*/}

Reviewer: ```suggestion
## Aggiornare gli array nello state {/*updating-arrays-in-state*/}
```

<LearnMore path="/learn/updating-arrays-in-state">

Leggi **[Aggiornare gli Array nello State](/learn/updating-arrays-in-state)** per imparare ad aggiornare gli array correttamente.

</LearnMore>

> ## Qual è il Prossimo Passo? {/*whats-next*/}

Reviewer: ```suggestion
## Qual è il prossimo passo? {/*whats-next*/}
```
